### PR TITLE
Interleave bug

### PR DESCRIPTION
**Pull request by pranavb-ca**
We now do not leave a buffer deinterleaved in storage when there are unaligned loads from or stores to it.

### CODE REVIEW

**Comment by pranavb-ca** [comment]
@aankit-ca, @dpalermo, @ronlieb, @suyogsarda : PTAL

File: src/HexagonOptimize.cpp [hunk 1]
```diff
@@ -1528,6 +1565,13 @@ class EliminateInterleaves : public IRMutator2 {
                 // which is only true if any of the stores are
                 // actually interleaved (and don't just yield an
                 // interleave).
+                internal_assert(aligned_buffer_access.contains(op->name) && "Buffer not found in scope");
+                bool &aligned_accesses = aligned_buffer_access.ref(op->name);
+                int aligned_offset = 0;
+
+                HexagonAlign alignment = alignment_analyzer.is_aligned(op, &aligned_offset);
```
Comment (dsharletg) [inline src/HexagonOptimize.cpp]: This probably should have a comment

File: src/HexagonAlignment.h [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+        this->alignment_info.set_containing_scope(&alignment_info);
+    }
+    /** Analyze the index of a load/store instruction for alignment
+     *  It returns HexagonAlign::Unknown, HexagonAlign::Unaligned or HexagonAlign::Aligned
+     */
+    template<typename T>
+    HexagonAlign is_aligned_impl(const T *op, int native_lanes, int *aligned_offset) {
+        debug(3) << "HexagonAlignmentAnalyzer: Check if " << op->index << " is aligned to a "
+                 << required_alignment << " byte boundary\n";
+        //        debug(3) << "Type: " << op->type << "\n";
```
Comment (dsharletg) [inline src/HexagonAlignment.h]: Remove commented code

File: src/HexagonOptimize.cpp [hunk 3]
```diff
@@ -1443,6 +1461,10 @@ class EliminateInterleaves : public IRMutator2 {
     };
     Scope<BufferState> buffers;
 
+    // True for buffers that have loads and stores that are aligned
```
Comment (dsharletg) [inline src/HexagonOptimize.cpp]: Maybe more correct to say

    // False for buffers that have any loads or stores that are unaligned

? (existing is a bit ambiguous: all loads or some loads?)

File: src/HexagonOptimize.cpp [hunk 4]
```diff
@@ -1500,8 +1529,16 @@ class EliminateInterleaves : public IRMutator2 {
                 // interleave itself, we don't want to change the
                 // buffer state.
             }
-        }
 
+            // Should we check this only when state == BufferState::Interleaved
```
Comment (dsharletg) [inline src/HexagonOptimize.cpp]: Is this a TODO? Or add a question mark?

**Comment by pranavb-ca** [comment]
@dsharletg - The perf tests we run returned with the same perf numbers (modulo some noise on device) before and after I made the alignment change. Only after that did i proceed to fixing the actual problem in EliminateInterleaves

File: src/AlignLoads.cpp [hunk 5]
```diff
@@ -66,30 +63,16 @@ class AlignLoads : public IRMutator2 {
             // non-constant strides.
             return IRMutator2::visit(op);
         }
-        int lanes = ramp->lanes;
-        int native_lanes = required_alignment / op->type.bytes();
 
-        if (!(*const_stride == 1 || *const_stride == 2 || *const_stride == 3)) {
```
Comment (dsharletg) [inline src/AlignLoads.cpp]: I think this logic has been lost, I think we'll rewrite loads even with huge constant strides whereas before we did not.

File: src/AlignLoads.cpp [hunk 6]
```diff
@@ -66,30 +63,16 @@ class AlignLoads : public IRMutator2 {
...
-            // If the ramp isn't stride 1, 2, or 3, don't handle it.
-
-            // TODO: We should handle reverse vector loads (stride ==
-            // -1), maybe others as well.
+        int aligned_offset = 0;
+        HexagonAlign alignment = alignment_analyzer.is_aligned(op, &aligned_offset);
+        if (alignment == HexagonAlign::Unknown) {
```
Comment (dsharletg) [inline src/AlignLoads.cpp]: I don't think is_aligned ever actually returns Unknown right now...

File: src/AlignLoads.cpp (hunk 5 above)
Comment (pranavb-ca) [reply src/AlignLoads.cpp]: Nice catch, thanks.

**Comment by pranavb-ca** [comment]
@dsharletg  - thanks for your comments. Addressed them. PTAL.

**Comment by steven-johnson** [comment]
I just updated to master to trigger the buildbots (now that they are finally running again).

**Comment by steven-johnson** [comment]
The one failure is a buildbot stall that I manually killed; since this is approved and everything else is clean, I'm going to land it.

**Pull request merged**